custom/icds_reports/management/commands/dashboard_usage_report.py

```diff
@@ -0,0 +1,156 @@
+            self.get_dashboard_usage_counts(start_date, end_date, domain)
+
+        username_state_mapping = self.get_tabular_data(usage_data, tab_usage_counts, tab_indicators_count)
+        print(f'Request data written to file {REQUEST_DATA_CACHE}')
+        self.get_cas_data(cas_counts, username_state_mapping)
+        print(f'Request data written to file {CAS_DATA_CACHE}')
+
+    def get_dashboard_usage_counts(self, start_date, end_date, domain):
+        """
+        :param start_date: start date of the filter
+        :param end_date: end date of the filter
+        :param domain
+        :return: returns the counts of no of downloads of each and total reports for  all usernames
+        """
+        print(f'Compiling usage counts for users')
+        tabular_user_counts = defaultdict(int)
+        cas_user_counts = defaultdict(int)
+        tabular_user_indicators = defaultdict(lambda: [0] * 10)
+
+        urls = ['/a/' + domain + '/cas_export', '/a/' + domain + '/icds_export_indicator']
+        query = list(
+                ICDSAuditEntryRecord.objects.filter(url__in=urls, time_of_use__gte=start_date,
+                                                    time_of_use__lte=end_date)
+                .annotate(indicator=Cast(KeyTextTransform('indicator', 'post_data'), IntegerField()))
+                .filter(indicator__lte=THR_REPORT_EXPORT).values('indicator', 'username', 'url')
+                .annotate(count=Count('indicator')).order_by('username', 'indicator'))
+        for record in query:
+            if record['url'] == urls[0]:
+                cas_user_counts[record['username'].split('@')[0]] += record['count']
+            else:
+                tabular_user_counts[record['username'].split('@')[0]] += record['count']
+                tabular_user_indicators[record['username'].split('@')[0]][int(record['indicator']) - 1]\
+                    = record['count']
+
+        return tabular_user_counts, tabular_user_indicators, cas_user_counts
+
+    def get_tabular_data(self, usage_data, tab_total_counts, tab_indicators_count):
+        tab_data = list()
+        extra_headers = ['Total', 'Child', 'Pregnant Women', 'Demographics', 'System Usage',
+                         'AWC infrastructure', 'Child Growth Monitoring List', 'ICDS - CAS Monthly Register',
+                         'AWW Performance Report', 'LS Performance Report', 'Take Home Ration']
+        username_vs_state_name = defaultdict()
+        print(f'Compiling request data for {len(usage_data)} users')
+        headers = usage_data[0][:7]
+        headers.extend(extra_headers)
+        usage_data = usage_data[1:]
+        for chunk in chunked(with_progress_bar(usage_data), 500):
+            for data in chunk:
+                username = data[4]
+                indicator_count = tab_indicators_count[username]
+                csv_row = data[:7]
+                csv_row.append(tab_total_counts[username])
+                csv_row.extend(indicator_count)
```

Comment:
How do we ensure that indicator_count is ordered in same way their corresponding headers are orders?

Reply:
we have ordered them on ascending order of indicator number which is same as that of headers

Reply:
I see